### PR TITLE
Paywall

### LINKED ISSUE #106: add paywall component
Need to port the bootstrap+jquery widget from dcrpayments.

### PR DESCRIPTION
**Pull request by vctt94**
fixes #106 

Still need some styling and a faucet button for automatic testing, as mentioned at #164 

Right now is needed to manually change the paywall address for testing.

### CODE REVIEW

**Comment by dajohi** [comment]
@vctt94 fix the test failures?

**Comment by vctt94** [comment]
Tests are failing because of https://github.com/decred/politeiagui/issues/147

File: src/selectors/api.js [hunk 1]
```diff
@@ -138,3 +138,8 @@ export const setStatusProposalIsRequesting = isApiRequestingSetStatusProposal;
 export const setStatusProposalToken = compose(get("token"), apiSetStatusProposalPayload);
 export const setStatusProposalError = apiSetStatusProposalError;
 export const redirectedFrom = get(["api", "login", "redirectedFrom"]);
+export const paywallAddress = compose(get("paywalladdress"), apiNewUserResponse);
+export const paywallAmount = compose(get("paywallamount"), apiNewUserResponse);
+export const VerificationToken = compose(get("verificationToken"), apiNewUserResponse);
```
Comment (sndurkin) [inline src/selectors/api.js]: The variable name here should use camelCase, and the string that's sent from Politeia will be all lowercase (same issue with the 2 lines below this).

File: src/connectors/paywall.js [hunk 2]
```diff
@@ -0,0 +1,17 @@
+import { connect } from "react-redux";
+import { bindActionCreators } from "redux";
+import * as sel from "../selectors";
+import * as act from "../actions";
+
+export default connect(
+  sel.selectorMap({
+    paywallAddress: sel.paywallAddress,
+    paywallAmount: sel.paywallAmount,
+    VerificationToken: sel.VerificationToken,
```
Comment (sndurkin) [inline src/connectors/paywall.js]: We should use camelCase here.

**Comment by vctt94** [comment]
Thanks for reviewing my PR.

I believe we can use in-line instead of a modal, should I remove the modal from the code? The way it is now, I believe it can be reusable for some modal we may have someday. 

I can change the polling for another place where it is always called, like the navbar. But I agree with you, it is better to wait for the server to inform it is paid.

But are we going to wait for the server to say it is paid and check for the mined blocks on the gui? Or will it be all on the server side and the gui only asks if it good to access the submit proposal area?

**Comment by sndurkin** [comment]
> But are we going to wait for the server to say it is paid and check for the mined blocks on the gui? Or will it be all on the server side and the gui only asks if it good to access the submit proposal area?

I think the server will reply with `hasPaid` on `/user/login` and `/user/me`, and if it's false, the GUI will poll for mined blocks, because it's possible that the user verifies his email, logs in, and then the transaction is confirmed on the blockchain he clicks around on the site. In that case, we will want to enable commenting and proposal submissions without the user having to press F5 to refresh the browser. Does that make sense?

**Comment by vctt94** [comment]
I removed the modal and made it a normal component, I created the hadPaid, which is get from login and me, and added the poll on the header if it hasPaid === false. But we have a problem, I may not have the paywall address and I am not sure if it makes sense to send it together with `/user/login` and `/user/me` route. What do you think?

File: mocks/api/v1/user/me/GET.json [hunk 3]
```diff
@@ -1,4 +1,5 @@
 {
-  "email": "foo@foo.com",
-  "isadmin": true
-}
+    "hasPaid": true,
```
Comment (sndurkin) [inline mocks/api/v1/user/me/GET.json]: This should be all lowercase.

File: src/lib/external_api.js [hunk 4]
```diff
@@ -0,0 +1,14 @@
+const dcrdataURL = (network, address) => `https://${network}.dcrdata.org/api/address/${address}/raw`;
+const getRawAddresses = (network, address) => fetch(dcrdataURL(network, address)).then(r => {
+  // work around when transactions are not paid and dcrdata api returns Unprocessable Entity
+  if(r.statusText === "Unprocessable Entity") {
+    return null;
+  }
+  return r.json();
+});
+
+export const getPaymentsByAddress = address => {
```
Comment (jolan) [inline src/lib/external_api.js]: Makes more sense to just check the first letter of the payment address. i.e. D = mainnet, T = testnet

File: src/components/Paywall/Page.js [hunk 5]
```diff
@@ -0,0 +1,27 @@
...
+  <div className="modal-wrapper">
+    <div className="paywall-header">
+      <h2 className="paywall-header-title">Paywall</h2>
+    </div>
+    <div className="paywall-content">
+      <p>Please send exaclty {props.paywallAmount}DCR to <br/>{props.paywallAddress}</p>
+    </div>
+    <div>
+      Status: {
+        grantAccess ?  <p>Approved</p> : <p>Waiting for blocks to get mined</p>
```
Comment (sndurkin) [inline src/components/Paywall/Page.js]: Let's call this "Payment received" and "Waiting for payment"

File: src/components/Paywall/Page.js [hunk 6]
```diff
@@ -0,0 +1,27 @@
...
+  onSubmit,
+  grantAccess,
+  ...props
+}) => (
+  <div className="modal-wrapper">
+    <div className="paywall-header">
+      <h2 className="paywall-header-title">Paywall</h2>
+    </div>
+    <div className="paywall-content">
+      <p>Please send exaclty {props.paywallAmount}DCR to <br/>{props.paywallAddress}</p>
```
Comment (sndurkin) [inline src/components/Paywall/Page.js]: Typo, should be "exactly".

File: src/components/Paywall/Page.js [hunk 7]
```diff
@@ -0,0 +1,27 @@
...
+    </div>
+    <div>
+      Status: {
+        grantAccess ?  <p>Approved</p> : <p>Waiting for blocks to get mined</p>
+      }
+    </div>
+    <div className="paywall-footer">
+      <a className="btn">Pay!</a>
```
Comment (sndurkin) [inline src/components/Paywall/Page.js]: I don't think a button is necessary here, because the user will pay with Decrediton, right?

File: src/components/Paywall/index.js [hunk 8]
```diff
@@ -0,0 +1,54 @@
...
+import PaywallPage from "./Page";
+import paywallConnector from "../../connectors/paywall";
+
+class Paywall extends React.Component {
+  constructor(props) {
+    super(props);
+    this.state = this.getInitialState();
+  }
+
+  getInitialState() {
```
Comment (tiagoalvesdulce) [inline src/components/Paywall/index.js]: getInitialState should not be used with ES6 classes, you can initialize the state in the constructor.

File: src/components/Paywall/index.js [hunk 9]
```diff
@@ -0,0 +1,54 @@
...
+        }}
+      />
+    );
+  }
+
+  resetState() {
+    this.setState(this.getInitialState());
+  }
+
+  onSubmit() {
```
Comment (tiagoalvesdulce) [inline src/components/Paywall/index.js]: onSubmit function is empty here.

**Comment by vctt94** [comment]
I fixed your comments and also I added a warningPaywallNotPaid component for not showing the submit proposal button and save button on comments.

File: src/components/snew/CommentForm.js [hunk 10]
```diff
@@ -77,18 +79,23 @@ const CommentForm = ({
...
-                className="cancel"
-                onClick={() => onSetReplyParent()}
-                type="button"
-              >
-              cancel
-              </button>)) || null}
-            </div>
+            {
+              !grantAccess ? <WarningPaywallNotPaid
+                message="You can not Comment on proposals before paying the paywall" /> : (
```
Comment (sndurkin) [inline src/components/snew/CommentForm.js]: 2 comments:

* Could you include the Save button but just disable it and the text area? I think it will make it a little clearer that this is the comment form if the button is present too.
* I'd like to refrain from calling it a paywall, can you rephrase it to: "You must pay the registration fee before you can comment on proposals"

File: src/components/snew/SubmitTextSidebox.js [hunk 11]
```diff
@@ -1,26 +1,30 @@
...
+            Submit a new Proposal
+          </Link>
+          <div className="nub" />
+        </div>
       </div>
     </div>
-  </div>
-) : null;
+  ) : <WarningPaywallNotPaid
+    message="You can not send Proposals before paying The Paywall" />
```
Comment (sndurkin) [inline src/components/snew/SubmitTextSidebox.js]: Same comments here:

* Could we show the "Submit a new proposal" button but disable it?
* Can you rephrase the message to: "You must pay the registration fee before you can submit a proposal."

File: src/actions/external_api.js [hunk 12]
```diff
@@ -0,0 +1,68 @@
...
+  return txid;
+};
+
+const checkTransaction = (transaction, addressToMatch, amount) => {
+  let addressSeen = false;
+  let addressValue = 0;
+  for (let voutData of transaction["vout"]) {
+    const addresses = voutData["scriptPubKey"]["addresses"];
+    if (addresses) {
+      addresses.forEach(address => {
```
Comment (tiagoalvesdulce) [inline src/actions/external_api.js]: This line is throwing a [no-loop-func](https://eslint.org/docs/rules/no-loop-func) warning. You should basically create a function outside the `for (let voutData ... )` loop and call it inside the `addresses.forEach`.

**Comment by vctt94** [comment]
Made the sean changes and I changed forEach, because it needs a callback function.

File: mocks/api/v1/user/me/GET.json [hunk 13]
```diff
@@ -1,5 +1,4 @@
 {
-    "email": "foo@foo.com",
```
Comment (tiagoalvesdulce) [inline mocks/api/v1/user/me/GET.json]: Removing this messes up with `yarn uidev` and Travis

**Comment by vctt94** [comment]
Ok, I am redirecting now and I returned the email to the mock api.

I believe we should not have tests depending on the mock API. If the test need to be logged in to pass, it needs to login. Maybe we should have an issue for that.
